src/backend/distributed/shared_library_init.c

```diff
@@ -2597,7 +2599,11 @@ static void
 ApplicationNameAssignHook(const char *newval, void *extra)
 {
 	ResetHideShardsDecision();
-	ResetCitusBackendType();
+	DetermineCitusBackendType(newval);
+	if (RanStartupCitusBackend)
```

Comment:
This is very much a hacky way of checking if we are after the startup phase that makes catalogs available. Because ApplicationNameAssignHook is called the first time so early that we cannot read the group_id catalog table.

Reply:
So far, the strategy was to only reset cache here, and then load the actual value later. That seems like a reasonable option to follow for GPID as well. Would that be feasible?

Reply:
I thought a bit about it, but afaict that's not feasible. The main reason is that the gpid is not just some variable that is local to the current backend. It's part of shared memory and other backends will want to read it, so it should be updated as soon as possible. This is also why we set the gpid and distributedCommandOriginator values in the auth hook already when possible.

So I'm starting to think my that my hacky solution is probably one of the best ones we can come up with. I'll add some more comments explaining this.

Reply:
makes sense

Reply:
Added a comment